File: ping.py
import requests
import datetime
import hashlib
import hmac
import urllib.parse
import string
import sys

##############################
# Configurações de ambiente #
##############################

HOST = "URL"  # Exemplo
URL_BASE = f"https://{HOST}"

# Ajuste com seu AK e SK reais:
ACCESS_KEY = "AK"
SECRET_KEY = "SK"

# Ajuste o path do endpoint conforme a doc/ambiente real:
HTTP_URI = "/rest/cmsapp/v1/ping"  # Exemplo: "/rest/cmsapp/v1/ping"


##############################
# Funções utilitárias
##############################

def utc_timestamp_iso8601():
    """
    Gera a data/hora em formato UTC no padrão yyyy-MM-dd'T'HH:mm:ss.SSS'Z'.
    Exemplo: 2024-10-31T14:06:52.123Z
    """
    now = datetime.datetime.utcnow()
    millis = int(now.microsecond / 1000)
    return now.strftime(f"%Y-%m-%dT%H:%M:%S.{millis:03d}Z")


def sha256_hex(key, msg):
    """
    Calcula HMAC-SHA256 em hexa, equivalente ao SignerUtils.sha256Hex do Java.
    """
    return hmac.new(key.encode("utf-8"), msg.encode("utf-8"), hashlib.sha256).hexdigest()


def path_normalize(value):
    """
    Faz a normalização estilo 'PathUtils.normalize' do Java:
      - Faz percent-encode de caracteres fora do conjunto unreserved
    """
    # urllib.parse.quote com safe="-._~" impede que esses caracteres sejam codificados
    return urllib.parse.quote(value, safe="-._~")


def build_signed_headers(headers):
    """
    Converte as chaves de cabeçalho para lowercase e retorna em novo dicionário.
    Importante: a key 'authorization' não entra no cálculo, e host é obrigatório.
    """
    if not headers or "host" not in {k.lower() for k in headers.keys()}:
        raise ValueError("Cabeçalho 'host' é obrigatório no signedHeaders.")
    new_headers = {}
    for k, v in headers.items():
        new_headers[k.lower()] = v.strip()
    return new_headers


def build_canonical_headers(signed_headers):
    """
    Monta a string CanonicalHeaders:
      - Cada linha: "<lowercase_header_name>:<normalized_header_value>"
      - Ordenado alfabeticamente
      - Separado por \n
    """
    sorted_header_names = sorted(signed_headers.keys())
    lines = []
    for name in sorted_header_names:
        normalized_name = path_normalize(name)
        normalized_value = path_normalize(signed_headers[name])
        lines.append(f"{normalized_name}:{normalized_value}")
    return "\n".join(lines)


def build_signed_headers_string(signed_headers):
    """
    Monta a string 'content-length;content-type;host'
    ordenada alfabeticamente por nome do header.
    """
    sorted_header_names = sorted(signed_headers.keys())
    return ";".join([name.lower() for name in sorted_header_names])


def build_canonical_request(method, uri, query_params, signed_headers, body):
    """
    Monta o CanonicalRequest no formato:
      HttpMethod \n
      HttpURI \n
      (HttpParameters) \n
      SignedHeaders(...) \n
      CanonicalHeaders(...) \n
      Normalize(body)
    """
    # Garante que o uri comece com '/'
    if not uri.startswith("/"):
        uri = "/" + uri

    # Monta query string
    canonical_query = ""
    if query_params:
        kvs = []
        for k, v in query_params.items():
            k_enc = path_normalize(k)
            v_enc = path_normalize(v) if v else ""
            kvs.append(f"{k_enc}={v_enc}")
        kvs.sort()
        canonical_query = "&".join(kvs)

    # SignedHeaders string
    signed_headers_str = build_signed_headers_string(signed_headers)
    # CanonicalHeaders string
    canonical_headers_str = build_canonical_headers(signed_headers)
    # Normaliza o body
    normalized_body = path_normalize(body) if body else ""

    parts = [
        method.upper(),
        uri,
        canonical_query,  # se vazio, fica ""
        signed_headers_str,
        canonical_headers_str,
        normalized_body,
    ]

    return "\n".join(parts)


def build_authorization_header(access_key, secret_key,
                               method, uri,
                               query_params, payload,
                               signed_headers):
    """
    Gera o header "Authorization" usando a regra:
      1) authStringPrefix = "auth-v2/AK/timestamp/signedHeaders"
      2) signingKey = sha256Hex(secretKey, authStringPrefix)
      3) signature = sha256Hex(signingKey, canonicalRequest)
      4) authorization = authStringPrefix + "/" + signature
    """
    ts = utc_timestamp_iso8601()

    auth_version = "auth-v2"
    signed_headers_str = build_signed_headers_string(signed_headers)
    auth_string_prefix = f"{auth_version}/{access_key}/{ts}/{signed_headers_str}"

    print("DEBUG >> authStringPrefix =", auth_string_prefix)

    signing_key = sha256_hex(secret_key, auth_string_prefix)
    print("DEBUG >> signingKey =", signing_key)

    canonical_request = build_canonical_request(
        method, uri, query_params, signed_headers, payload
    )
    print("DEBUG >> CanonicalRequest:\n", canonical_request)

    signature = sha256_hex(signing_key, canonical_request)
    print("DEBUG >> Signature =", signature)

    authorization = f"{auth_string_prefix}/{signature}"
    return authorization, ts


##############################
# Exemplo de uso (POST)
##############################

def cms_ping_test():
    """
    Tenta enviar POST para o CMS (ou API) no endpoint HTTP_URI.
    Se der erro, printa status code e response pra ajudar no debug.
    """
    # Montar URL final
    full_url = f"{URL_BASE}{HTTP_URI}"
    print(f"DEBUG >> Full URL = {full_url}")

    # Montar body
    body = '{"say": "Hello world!"}'
    content_length = str(len(body.encode("utf-8")))

    # Cabeçalhos obrigatórios
    headers_inicial = {
        "Host": HOST,
        "Content-Length": content_length,
        "Content-Type": "application/json;charset=UTF-8",
    }

    # Transformar pra signed_headers
    signed_headers = build_signed_headers(headers_inicial)

    print("\n==== [DEBUG] HEADERS INICIAIS ====")
    for k, v in headers_inicial.items():
        print(f"{k}: {v}")
    print("==================================")

    print("\n==== [DEBUG] ACCESS_KEY / SECRET_KEY ====")
    print("ACCESS_KEY:", ACCESS_KEY)
    print("SECRET_KEY:", SECRET_KEY)
    print("=========================================\n")

    # Gera o Authorization
    auth_header, timestamp = build_authorization_header(
        ACCESS_KEY,
        SECRET_KEY,
        method="POST",
        uri=HTTP_URI,
        query_params=None,  # se houver params, passe em dict
        payload=body,
        signed_headers=signed_headers
    )

    final_headers = {
        **headers_inicial,
        "Authorization": auth_header
    }

    print("\n==== [DEBUG] HEADERS FINAIS ENVIADOS ====")
    for k, v in final_headers.items():
        print(f"{k}: {v}")
    print("=========================================")

    # Envia a requisição
    try:
        # Caso tenha problema de certificado SSL, `verify=False` ignora.
        resp = requests.post(full_url, data=body, headers=final_headers, verify=False)
    except requests.RequestException as e:
        print("ERRO >> Falha na requisição HTTP:", e)
        sys.exit(1)

    print("\n==== [DEBUG] RESPOSTA ====")
    print("HTTP Status Code:", resp.status_code)
    print("Response Body:", resp.text)
    print("Response Headers:", resp.headers)
    print("==========================")

    # Dicas de troubleshooting
    if resp.status_code == 404:
        print("DICA: 404 significa que o endpoint provavelmente não existe.")
        print(f"      Verifique se {HTTP_URI} está correto.")
    elif resp.status_code == 401:
        print("DICA: 401 significa que as credenciais ou a assinatura podem estar incorretas.")
        print("      Verifique seu AccessKey / SecretKey ou compare a doc da API.")
    elif resp.status_code == 403:
        print("DICA: 403 normalmente é 'Forbidden'. Pode ser permissão ou assinatura incorreta.")
    elif 400 <= resp.status_code < 500:
        print("DICA: Algum problema do lado do cliente. Consulte a doc para ver se falta algo.")
    elif resp.status_code >= 500:
        print("DICA: Erro do lado do servidor (5xx). O serviço pode estar indisponível ou com bug.")


def main():
    cms_ping_test()


if __name__ == "__main__":
    main()

File: ping.go
package main

import (
    "crypto/hmac"
    "crypto/sha256"
    "crypto/tls"
    "encoding/hex"
    "fmt"
    "io"
    "net/http"
    "net/url"
    "sort"
    "strings"
    "time"
)

// Configurações de ambiente
const (
    Host      = "10.22.26.181:28080"      // IP:Port
    AccessKey = "globalaktest"           // Seu AK
    SecretKey = "1q************20"       // Seu SK
    Scheme    = "https"
)

// Gera timestamp em UTC no formato yyyy-MM-dd'T'HH:mm:ss.SSS'Z'
func utcTimestampISO8601() string {
    now := time.Now().UTC()
    millis := now.Nanosecond() / 1_000_000
    return fmt.Sprintf("%04d-%02d-%02dT%02d:%02d:%02d.%03dZ",
        now.Year(), now.Month(), now.Day(),
        now.Hour(), now.Minute(), now.Second(), millis)
}

// Implementação estilo sha256Hex(secret, message)
func sha256Hex(key, data string) string {
    mac := hmac.New(sha256.New, []byte(key))
    mac.Write([]byte(data))
    return hex.EncodeToString(mac.Sum(nil))
}

// Faz o “percent-encode” estilo PathUtils.normalize()
func pathNormalize(value string) string {
    // url.QueryEscape codifica demais (por exemplo, `~` e outros).
    // Para ficar mais parecido com a doc, podemos usar url.PathEscape, 
    // mas esse também difere um pouco. Então vamos manualmente:
    return customPercentEncode(value)
}

// Precisamos de um encode mais customizado (similar ao Java), 
// mas pra simplificar vou fazer algo “próximo” usando substituição manual.
func customPercentEncode(value string) string {
    // Vou fazer um strings.Builder e “escapar” apenas o que não está em [A-Za-z0-9-_.~].
    // É um encode básico. Em produção, você pode refinar caso precise ficar idêntico.
    var sb strings.Builder
    for _, c := range value {
        // Verifica se é caractere “unreserved”
        if (c >= 'A' && c <= 'Z') ||
            (c >= 'a' && c <= 'z') ||
            (c >= '0' && c <= '9') ||
            c == '-' || c == '.' || c == '_' || c == '~' {
            sb.WriteRune(c)
        } else {
            // Percent-encode
            sb.WriteString(fmt.Sprintf("%%%02X", c))
        }
    }
    return sb.String()
}

// Monta a lista “header1;header2;header3” com nomes em minúsculo ordenados
func buildSignedHeadersString(signedHeaders map[string]string) string {
    var keys []string
    for k := range signedHeaders {
        keys = append(keys, strings.ToLower(k))
    }
    sort.Strings(keys)
    return strings.Join(keys, ";")
}

// Monta o CanonicalHeaders:
//  headerName:headerValue\n ...
//  ordenados alfabeticamente
//  com name e value normalizados
func buildCanonicalHeaders(signedHeaders map[string]string) string {
    var lines []string
    // Ordena pelas chaves
    var keys []string
    for k := range signedHeaders {
        keys = append(keys, strings.ToLower(k))
    }
    sort.Strings(keys)

    for _, k := range keys {
        // Normaliza a key e o value
        normKey := pathNormalize(k)
        normVal := pathNormalize(signedHeaders[k])
        lines = append(lines, fmt.Sprintf("%s:%s", normKey, normVal))
    }
    // Junta com \n
    return strings.Join(lines, "\n")
}

// Monta a query canonical (key=value&key2=value2...), ordenada
func buildCanonicalQueryString(queryParams map[string]string) string {
    if len(queryParams) == 0 {
        return ""
    }
    var kvs []string
    for k, v := range queryParams {
        normK := pathNormalize(k)
        normV := pathNormalize(v)
        kvs = append(kvs, fmt.Sprintf("%s=%s", normK, normV))
    }
    sort.Strings(kvs)
    return strings.Join(kvs, "&")
}

// Monta o canonical request:
//  HttpMethod \n
//  HttpURI \n
//  (HttpParameters?) \n
//  SignedHeaders(...) \n
//  CanonicalHeaders(...) \n
//  Normalize(body)
func buildCanonicalRequest(method, uri string, queryParams map[string]string, signedHeaders map[string]string, body string) string {
    // Garante que uri comece com /
    if !strings.HasPrefix(uri, "/") {
        uri = "/" + uri
    }

    canonicalQuery := buildCanonicalQueryString(queryParams)
    signedHeadersStr := buildSignedHeadersString(signedHeaders)
    canonicalHeadersStr := buildCanonicalHeaders(signedHeaders)
    normalizedBody := ""
    if body != "" {
        normalizedBody = pathNormalize(body)
    }

    // Monta as linhas
    // Se queryParams existir, canonicalQuery. Se não, string vazia
    // Observação: a doc coloca a queryParams + "\n" mesmo se estiver vazia (outra linha)
    parts := []string{
        strings.ToUpper(method),
        uri,
        canonicalQuery,       // se for vazio, fica string vazia
        signedHeadersStr,
        canonicalHeadersStr,
        normalizedBody,
    }

    return strings.Join(parts, "\n")
}

func buildAuthorizationHeader(accessKey, secretKey, method, uri string,
    queryParams map[string]string,
    body string,
    signedHeaders map[string]string) (string, string) {

    timestamp := utcTimestampISO8601()
    // authStringPrefix = "auth-v2/AK/timestamp/signedHeaders"
    authVersion := "auth-v2"
    shStr := buildSignedHeadersString(signedHeaders)
    authStringPrefix := fmt.Sprintf("%s/%s/%s/%s", authVersion, accessKey, timestamp, shStr)

    // signingKey = sha256Hex(secretKey, authStringPrefix)
    signingKey := sha256Hex(secretKey, authStringPrefix)

    // canonicalRequest
    canonicalRequest := buildCanonicalRequest(method, uri, queryParams, signedHeaders, body)

    // signature = sha256Hex(signingKey, canonicalRequest)
    signature := sha256Hex(signingKey, canonicalRequest)

    // Authorization = authStringPrefix + "/" + signature
    authorization := fmt.Sprintf("%s/%s", authStringPrefix, signature)
    return authorization, timestamp
}

// Exemplo de teste do CMS / ping
func cmsPingTest() error {
    // Endpoint
    httpURI := "/rest/cmsapp/v1/ping" // ex: "/rest/cmsapp/v1/ping"
    fullURL := fmt.Sprintf("%s://%s%s", Scheme, Host, httpURI)

    // Body
    body := `{"say":"Hello world!"}`
    contentLength := fmt.Sprintf("%d", len([]byte(body)))

    // Cabeçalhos iniciais
    headers := map[string]string{
        "Host":           Host,
        "Content-Length": contentLength,
        "Content-Type":   "application/json;charset=UTF-8",
    }

    // Gera Authorization
    authHeader, timestamp := buildAuthorizationHeader(
        AccessKey, SecretKey,
        "POST", httpURI,
        nil, // sem query params nesse ping
        body,
        headers,
    )

    // Monta a requisição
    req, err := http.NewRequest("POST", fullURL, strings.NewReader(body))
    if err != nil {
        return err
    }

    // Seta cabeçalhos
    for k, v := range headers {
        req.Header.Set(k, v)
    }
    // Adiciona o Authorization
    req.Header.Set("Authorization", authHeader)

    // Se precisar ignorar certificado SSL “inseguro”, crie um *http.Client custom:
    tr := &http.Transport{
        TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
    }
    client := &http.Client{Transport: tr}

    // Executa
    resp, err := client.Do(req)
    if err != nil {
        return err
    }
    defer resp.Body.Close()

    // Lê resposta
    bodyBytes, err := io.ReadAll(resp.Body)
    if err != nil {
        return err
    }

    fmt.Println("Status:", resp.StatusCode)
    fmt.Println("Timestamp que enviamos:", timestamp)
    fmt.Println("Authorization que enviamos:", authHeader)
    fmt.Println("Resposta do servidor:", string(bodyBytes))
    return nil
}

func main() {
    err := cmsPingTest()
    if err != nil {
        fmt.Println("Erro ao chamar cmsPingTest:", err)
    }
}
